utils/pattern/src/num_pattern/mod.rs

```diff
@@ -0,0 +1,366 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::fmt::{self, Write};
+use writeable::{PartsWrite, Writeable};
+
+#[cfg(feature = "alloc")]
+use crate::parser::{Parser, ParserOptions, PatternToken};
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+/// A string pattern with simple numeric placeholders, like `"{0} and {1}"`
+///
+/// # Backing Store
+///
+/// The data structure has a flexible backing data store. The only requirement for most
+/// functionality is that it implement `AsRef<str>`.
+///
+/// # Pattern String Encoding Details
+///
+/// The pattern is stored as a string with the following encoding:
+///
+/// - The first code point is the byte offset of the start of the string literals.
+/// - Code points 1 through N are the byte offsets of those placeholders.
+/// - The remainder of the string contains the string literals.
+///
+/// For example, a string with one placeholder, like "A{0}B", is encoded as: `"\x02\x03AB"`
```

Comment:
if you stored placeholder offsets from the beginning of the string we'd have a bit more range